unixfs/tar/writer.go

```diff
@@ -32,16 +33,30 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
 
 	// construct the tar writer
-	w, err := NewWriter(bufw, dag, compression)
+	w, err := NewWriter(ctx, dag, archive, compression, bufw)
 	if err != nil {
 		return nil, err
 	}
 
 	// write all the nodes recursively
 	go func() {
-		if err := w.WriteNode(ctx, nd, filename); err != nil {
-			pipew.CloseWithError(err)
-			return
+		if !archive && compression != gzip.NoCompression {
+			// the case when the node is a file
+			dagr, err := uio.NewDagReader(w.ctx, nd, w.Dag)
+			if err != nil {
+				pipew.CloseWithError(err)
+				return
+			}
+
+			if _, err := dagr.WriteTo(w.w); err != nil {
```

Comment:
yeah gziping is pseudo archive. it could sort of work... i think it's almost there, maybe take the function i gave for now and move on to other things. it's already cleaner